### PR TITLE
ocamllex: Accept line directives ended with CR/LF. (#12146)

### LINKED ISSUE #12146: ocamllex should accept #line-directive ended with CR/LF
`ocamllex` refuses line-directives ended with CR/LF like the below.

```
$ cat -e tmp.mll
# 1 "tmp.mll.cppo"^M$
^M$
rule token = parse^M$
  | _ { }^M$

$ ocamllex tmp.mll
File "tmp.mll", line 1, character 0: syntax error.
```

Use case: I'm building a tokenizer from CPPO-preprocessed source. On Windows, CPPO (v1.6.9) converts all LF to CF/LF. Because of that, `ocamllex` syntax errors occur on every line-directives generated by CPPO.

Looking for lex/lexer.mll and history, It seems that all CRs are ignored except for those at the end of line-directives. Maybe we need to ignore this case also?

### PR DESCRIPTION
**Pull request by scjung**
This PR fixes ocamllex to accept line-directives ended with CR/LF.

Use case: I'm building a tokenizer from CPPO-preprocessed source. On Windows, CPPO (v1.6.9) converts all LF to CF/LF. Because of that, ocamllex syntax errors occur on every line-directives generated by CPPO.

Looking at lex/lexer.mll and history, It seems that all CRs are ignored except for those at the end of line-directives.

### CODE REVIEW

File: lex/lexer.mll [hunk 1]
```diff
@@ -137,7 +137,7 @@ rule main = parse
       main lexbuf }
   | "#" [' ' '\t']* (['0'-'9']+ as num) [' ' '\t']*
     ('\"' ([^ '\010' '\013' '\"']* as name) '\"')?
-    [^ '\010' '\013']* '\010'
+    [^ '\010' '\013']* '\013'? '\010'
```
Comment (nojb) [inline lex/lexer.mll]: If we want to ignore _all_ CRs, then it would be better as follows:
```suggestion
    [^ '\010' '\013']* '\013'* '\010'
```

**Review by nojb — COMMENTED** [review]
PR looks reasonable to me. I left a suggestion.

**Comment by hhugo** [comment]
The change looks good to me. I think you should also open an issue on cppo. I think cppo should just preserve existing line ending

**Comment by nojb** [comment]
Also, a Changes entry is in order.

File: lex/lexer.mll (hunk 1 above)
Comment (scjung) [reply lex/lexer.mll]: You're right. will be fixed too.

File: Changes [hunk 2]
```diff
@@ -325,6 +325,9 @@ Working version
   used in any context where a character set is expected.
   (Nicolás Ojeda Bär, Martin Jambon, review by Sébastien Hinderer)
 
+- #12147: ocamllex: Allow carriage returns at the end of line directives.
+  (SeungCheol Jung)
```
Comment (nojb) [inline Changes]: Please add reviewers.

**Comment by damiendoligez** [comment]
CI fails because checking out `crlf_line_directive.mll` changes your CR-CR-LF line into a CR-LF in the test file and `check-labelled-interfaces.sh` then sees a diff that shouldn't be there. You probably need to declare the file `binary` in `.gitattributes`.

File: lex/lexer.mll [hunk 3]
```diff
@@ -137,7 +137,7 @@ rule main = parse
       main lexbuf }
   | "#" [' ' '\t']* (['0'-'9']+ as num) [' ' '\t']*
     ('\"' ([^ '\010' '\013' '\"']* as name) '\"')?
-    [^ '\010' '\013']* '\010'
+    [^ '\010' '\013']* '\013'* '\010'
```
Comment (damiendoligez) [inline lex/lexer.mll:140]: You could also do it this way:
```suggestion
    [^ '\010' ]* '\010'
```

**Review by damiendoligez — APPROVED** [review]
Looks good. Fix the CI problem and we can merge.

**Comment by gasche** [comment]
The CI is configured to complain about cr/lf at the end of files, and git does not handle it well either. The simplest route to merge this PR would be to simply remove test. @scjung, could you do this?

**Comment by hhugo** [comment]
> The change looks good to me. I think you should also open an issue on cppo. I think cppo should just preserve existing line ending

Has this been reported on the cppo project ?

**Comment by scjung** [comment]
> The CI is configured to complain about cr/lf at the end of files, and git does not handle it well either. The simplest route to merge this PR would be to simply remove test. @scjung, could you do this?

Sorry for late reply. I removed the test.

**Pull request merged**